src/common/eeprom.cpp

```diff
@@ -409,63 +413,115 @@ static eeprom_vars_t &eeprom_startup_vars() {
-    uint16_t addr;
-    uint16_t size;
-    uint16_t data_size;
-    void *data_ptr;
     if (id < EEPROM_VARCOUNT) {
+        if (variant8_get_type(var) == eeprom_map[id].type) {
+            eeprom_set_var(id, variant8_data_ptr(&var), eeprom_var_size(id));
+        } else {
+            // TODO: error
+            log_error(EEPROM, "%s: variant type missmatch on id: %x", __FUNCTION__, id);
+        }
+
+    } else {
+        assert(0 /* EEProm var Id out of range */);
+    }
+}
+
+/**
+ * @brief function that writes data to eeprom, also actualizes crc and RAM structure
+ *
+ * @param id eeprom record index
+ * @param var_ptr pointer to variable to be written
+ * @param var_size size of variable
+ */
+static void eeprom_set_var(enum eevar_id id, void *var_ptr, size_t var_size) {
+    eevar_id sheet_id = static_cast<enum eevar_id>(0);
+    eeprom_vars_t &vars = eeprom_startup_vars();
+    uint16_t var_eeprom_addr;
+    uint16_t sheet_eeprom_addr;
+    void *var_ram_addr;
+    void *sheet_ram_addr;
+    if (id < EEPROM_VARCOUNT) {
+        // Z offset is written to currnet sheet too
+        //TODO remove EEVAR_ZOFFSET to store it only once
 #if (EEPROM_FEATURES & EEPROM_FEATURE_SHEETS)
-        if (id == EEVAR_ZOFFSET && variant8_get_type(var) == VARIANT8_FLT) {
+        if (id == EEVAR_ZOFFSET) {
             variant8_t recent_sheet = eeprom_get_var(EEVAR_ACTIVE_SHEET);
             uint8_t index = variant8_get_ui8(recent_sheet);
-            uint16_t profile_address = eeprom_var_addr(static_cast<enum eevar_id>(EEVAR_SHEET_PROFILE0 + index));
-            float z_offset = variant8_get_flt(var);
-            st25dv64k_user_write_bytes(profile_address + MAX_SHEET_NAME_LENGTH, &z_offset, sizeof(float));
+            eevar_id sheet_id = static_cast<enum eevar_id>(EEVAR_SHEET_PROFILE0 + index);
+            //eeprom area
+            sheet_eeprom_addr = eeprom_var_addr(sheet_id);
+
+            //RAM area
+            sheet_ram_addr = eeprom_var_ptr(sheet_id, vars);
         }
 #endif
-        eeprom_lock();
-        if (variant8_get_type(var) == eeprom_map[id].type) {
-            size = eeprom_var_size(id);
-            data_size = variant8_data_size(&var);
-            if ((size == data_size) || ((variant8_get_type(var) == VARIANT8_PCHAR) && (data_size <= size))) {
-                addr = eeprom_var_addr(id);
-                data_ptr = variant8_data_ptr(&var);
-                st25dv64k_user_write_bytes(addr, data_ptr, data_size);
-                eeprom_update_crc32();
-            } else {
-                // TODO: error
-                log_error(EEPROM, "%s: invalid data size", __FUNCTION__);
+        size_t size = eeprom_var_size(id);
+        if (var_size <= size) {
+            var_eeprom_addr = eeprom_var_addr(id);
+            var_ram_addr = eeprom_var_ptr(id, vars);
+            //critical section
+            eeprom_lock();
+            if (sheet_id != static_cast<enum eevar_id>(0)) {
```

Comment:
And here you check if it is not zero. It can not be zero, as it is initialized to 0 and never used.